### PR TITLE
feat: fix blocking user 2 times

### PR DESCRIPTION
**Pull request by sophieur444**
## Summary

This PR fixes a critical bug where users could be blocked twice due to race conditions when rapidly toggling block/unblock actions. Additionally, it adds offline support for blocked user names and significantly improves test coverage for blocking/unblocking functionality.

The PR includes 706 lines to fix the bug, 1112 lines of tests and 246 deletions

This PR solves issue https://github.com/orgs/swent25-team-18/projects/3/views/1?pane=issue&itemId=144698402&issue=swent25-team-18%7Cmy-swiss-dorm%7C312

## Problem

- **Race Condition Bug**: Rapidly clicking block/unblock buttons could result in users being blocked multiple times or inconsistent UI state
- **Offline Support**: Blocked user names were not available when offline, requiring network access to display blocked contacts
- **Test Coverage**: Missing comprehensive tests for blocking/unblocking functionality, especially edge cases

### 1. Race Condition Fix (ViewUserProfileScreenViewModel)

Implemented a toggle token mechanism to prevent race conditions:

- **Toggle Token System**: Each block/unblock operation gets a unique token to identify the latest intent
- **Job Cancellation**: Previous in-flight operations are cancelled when a new toggle is initiated
- **Optimistic Updates**: UI updates immediately for better UX, with rollback on error
- **Late Result Ignoring**: Old async operations (like photo loading) are ignored if a newer toggle occurred

**Key Changes:**
- Added `toggleToken` counter to track the latest operation
- Cancel `toggleJob` and `photoJob` on each toggle
- Only apply state updates if the operation token matches the current token
- Separate photo loading job that respects the toggle token

### 2. Offline Support for Blocked User Names (SettingsViewModel)

Enhanced blocked contacts display to work offline:

- **Stored Names Priority**: Uses locally stored display names first (works offline)
- **Remote Fallback**: Fetches from remote if stored name not available
- **UID Fallback**: Uses UID as display name if profile name is blank
- **Graceful Error Handling**: Continues to work even if profile fetch fails

**Key Changes:**
- Added `getBlockedUserNames()` method to ProfileRepository interface
- Updated `SettingsViewModel.refresh()` to check stored names first
- Implemented three-tier fallback: stored name → remote fetch → UID

### 3. Comprehensive Test Coverage

Added extensive test coverage for blocking/unblocking functionality:

#### SettingsViewModel Tests (5 new tests):
- `refresh_usesStoredBlockedUserName_whenAvailable()` - Tests offline name retrieval
- `refresh_fetchesRemoteName_whenStoredNameNotAvailable()` - Tests remote fallback
- `refresh_fallsBackToUid_whenNameIsBlank()` - Tests UID fallback
- `refresh_usesStoredNameOverRemote_whenBothAvailable()` - Tests stored name priority
- `unblockUser_success_removesUserFromBlockedList()` - Tests successful unblocking

#### ViewUserProfileScreenViewModel Tests (6 new tests):
- `blockUser_optimisticUpdate_setsBlockedImmediately()` - Tests optimistic UI updates
- `blockUser_hidesProfilePicture_whenBlocked()` - Tests profile picture hiding
- `unblockUser_loadsProfilePicture_afterUnblock()` - Tests photo loading after unblock
- `blockUser_rapidToggle_preventsRaceCondition()` - Tests race condition prevention
- `blockUser_error_revertsOptimisticUpdate()` - Tests error handling and rollback

#### Test Fixes:
- Fixed async timing issues in existing tests
- Added `waitForRepositoryBlockedState()` helper for proper async waiting
- Improved error messages for better debugging

### Files Changed

**Core Implementation:**
- `ViewUserProfileScreenViewModel.kt` - Added toggle token mechanism
- `SettingsViewModel.kt` - Added offline name support with fallback logic
- `ProfileRepository.kt` - Added `getBlockedUserNames()` interface method
- `ProfileRepositoryFirestore.kt` - Implemented `getBlockedUserNames()` (returns empty map)
- `ProfileRepositoryHybrid.kt` - Implemented `getBlockedUserNames()` with local/remote sync
- `ProfileRepositoryLocal.kt` - Implemented `getBlockedUserNames()` for local storage

**Tests:**
- `SettingsViewModelTest.kt` - Added 5 new tests, fixed existing tests
- `ViewUserProfileScreenBlockedTest.kt` - Added 6 new tests, fixed async timing issues

### Key Implementation Patterns

1. **Toggle Token Pattern**: Prevents race conditions by tracking operation sequence
2. **Optimistic Updates**: Immediate UI feedback with error rollback
3. **Offline-First**: Prioritizes local data, falls back to remote when needed
4. **Graceful Degradation**: Multiple fallback layers ensure functionality always works

## Impact

- **Bug Fix**: Prevents users from being blocked multiple times
- **UX Improvement**: Blocked contacts visible offline
- **Code Quality**: Comprehensive test coverage ensures reliability
- **Maintainability**: Clear patterns for handling async operations and race conditions

## Related Issues

- Fixes race condition when rapidly toggling block/unblock
- Improves offline experience for blocked contacts management
- Addresses test coverage gaps for blocking functionality

disclaimer: this pr was done with the help and suggestions of ai

### CODE REVIEW

File: app/src/main/java/com/android/mySwissDorm/ui/profile/ViewUserProfileScreenViewModel.kt [hunk 1]
```diff
@@ -143,8 +156,10 @@ class ViewProfileScreenViewModel(
    * Blocks a user by adding their UID to the current user's blocked list in Firestore. Updates the
    * UI state to reflect the blocked status.
    *
-   * @param targetUid The UID of the user to block
-   * @param onError Callback invoked when blocking fails, receives error message
+   * Fixes rapid/repeated block<->unblock toggles by:
+   * - canceling any in-flight toggle/photo work on each click
+   * - optimistic UI update
+   * - ignoring late results using a token (so old unblock photo loads can't overwrite a new block)
```
Comment (rami-mallah) [inline app/src/main/java/com/android/mySwissDorm/ui/profile/ViewUserProfileScreenViewModel.kt]: Be sure to keep the `@param` for the doc.

File: app/src/main/java/com/android/mySwissDorm/ui/profile/ViewUserProfileScreenViewModel.kt [hunk 2]
```diff
@@ -153,26 +168,40 @@ class ViewProfileScreenViewModel(
...
    * Unblocks a user by removing their UID from the current user's blocked list in Firestore.
    * Updates the UI state accordingly.
    *
-   * @param targetUid The UID of the user to unblock.
-   * @param onError Callback invoked when unblocking fails, receives an error message.
+   * Fixes rapid/repeated block<->unblock toggles by:
+   * - canceling any in-flight toggle/photo work on each click
+   * - optimistic UI update
+   * - loading photo in a separate cancellable job
+   * - ignoring late results using a token
```
Comment (rami-mallah) [inline app/src/main/java/com/android/mySwissDorm/ui/profile/ViewUserProfileScreenViewModel.kt]: Same thing here for the `@param`

File: app/src/main/java/com/android/mySwissDorm/ui/profile/ViewUserProfileScreenViewModel.kt
Comment (rami-mallah) [inline app/src/main/java/com/android/mySwissDorm/ui/profile/ViewUserProfileScreenViewModel.kt]: Changes LGTM here, good job. Just pay attention to the doc.

**Review by rami-mallah — APPROVED** [review]
Thanks, that's good!
